### PR TITLE
change the method of initializing slice

### PR DESCRIPTION
**Pull request by sivchari**
- [x] Do only one thing
- [x] Non breaking API changes
- [x] Tested

### What did this pull request do?

Hi, I'm sorry for the trouble I caused you earlier. I'll be careful next time.
This is a small change, but it will send a PR.
There are places where the slice capacity is not specified.
I thought this would be better for Go's slice specification, since the capacity of the slice to be appended can be allocated in advance by the length.

### CODE REVIEW

**Comment by sivchari** [comment]
Oops, the run all tests in the local tests directory went through.
```shell
testing sqlite...
ok      gorm.io/gorm    0.845s
?       gorm.io/gorm/callbacks  [no test files]
ok      gorm.io/gorm/clause     2.513s
ok      gorm.io/gorm/logger     0.851s
?       gorm.io/gorm/migrator   [no test files]
ok      gorm.io/gorm/schema     1.325s
ok      gorm.io/gorm/utils      0.909s
?       gorm.io/gorm/utils/tests        [no test files]
go: finding module for package gorm.io/driver/postgres
go: finding module for package gorm.io/driver/mysql
go: finding module for package gorm.io/driver/sqlite
go: finding module for package gorm.io/driver/sqlserver
go: found gorm.io/driver/mysql in gorm.io/driver/mysql v1.0.4
go: found gorm.io/driver/postgres in gorm.io/driver/postgres v1.0.8
go: found gorm.io/driver/sqlite in gorm.io/driver/sqlite v1.1.4
go: found gorm.io/driver/sqlserver in gorm.io/driver/sqlserver v1.0.6
ok      gorm.io/gorm/tests      8.099s
testing mysql...
ok      gorm.io/gorm    0.339s
?       gorm.io/gorm/callbacks  [no test files]
ok      gorm.io/gorm/clause     1.449s
ok      gorm.io/gorm/logger     0.135s
?       gorm.io/gorm/migrator   [no test files]
ok      gorm.io/gorm/schema     0.559s
ok      gorm.io/gorm/utils      0.214s
?       gorm.io/gorm/utils/tests        [no test files]
ok      gorm.io/gorm/tests      44.006s
testing postgres...
ok      gorm.io/gorm    0.172s
?       gorm.io/gorm/callbacks  [no test files]
ok      gorm.io/gorm/clause     1.218s
ok      gorm.io/gorm/logger     0.181s
?       gorm.io/gorm/migrator   [no test files]
ok      gorm.io/gorm/schema     0.554s
ok      gorm.io/gorm/utils      0.134s
?       gorm.io/gorm/utils/tests        [no test files]
ok      gorm.io/gorm/tests      22.010s
testing sqlserver...
ok      gorm.io/gorm    0.183s
?       gorm.io/gorm/callbacks  [no test files]
ok      gorm.io/gorm/clause     1.629s
ok      gorm.io/gorm/logger     0.153s
?       gorm.io/gorm/migrator   [no test files]
ok      gorm.io/gorm/schema     0.815s
ok      gorm.io/gorm/utils      0.135s
?       gorm.io/gorm/utils/tests        [no test files]
ok      gorm.io/gorm/tests      28.078s
```

File: callbacks/associations.go [hunk 1]
```diff
@@ -340,10 +340,11 @@ func onConflictOption(stmt *gorm.Statement, s *schema.Schema, selectColumns map[
 
 func saveAssociations(db *gorm.DB, rel *schema.Relationship, values interface{}, selectColumns map[string]bool, restricted bool, defaultUpdatingColumns []string) error {
 	var (
-		selects, omits []string
-		onConflict     = onConflictOption(db.Statement, rel.FieldSchema, selectColumns, restricted, defaultUpdatingColumns)
-		refName        = rel.Name + "."
+		onConflict = onConflictOption(db.Statement, rel.FieldSchema, selectColumns, restricted, defaultUpdatingColumns)
+		refName    = rel.Name + "."
 	)
+	selects := make([]string, 0, len(selectColumns))
```
Comment (jinzhu) [inline callbacks/associations.go:346]: better to keep those codes in the `var` group?

Comment (sivchari) [reply callbacks/associations.go:346]: > better to keep those codes in the `var` group?

Thanks for the review!
You're right, I will put it back.

File: callbacks/associations.go [hunk 2]
```diff
@@ -39,10 +39,10 @@ func SaveBeforeAssociations(create bool) func(db *gorm.DB) {
 				switch db.Statement.ReflectValue.Kind() {
 				case reflect.Slice, reflect.Array:
 					var (
-						objs      []reflect.Value
 						fieldType = rel.Field.FieldType
 						isPtr     = fieldType.Kind() == reflect.Ptr
 					)
+					objs := make([]reflect.Value, 0, db.Statement.ReflectValue.Len())
```
Comment (jinzhu) [inline callbacks/associations.go:45]: change to following code?
```go
	var (
 						objs = make([]reflect.Value, 0, db.Statement.ReflectValue.Len())
 						fieldType = rel.Field.FieldType
 						isPtr     = fieldType.Kind() == reflect.Ptr
 	)
```

there are some similar changes could make

**Pull request merged**